src/machine.rs

```diff
@@ -1447,6 +1471,34 @@ async fn send_user_identities_to_callback(
     }
 }
 
+// helper for register_device_updated_callback: passes the user IDs into
+// the javascript function
+async fn send_device_updates_to_callback(
+    callback: &Function,
+    (_, device_updates): (IdentityChanges, DeviceChanges),
+) {
+    // get the user IDs of all the devices that have changed
+    let updated_chain = device_updates
+        .new
+        .into_iter()
+        .chain(device_updates.changed.into_iter())
+        .chain(device_updates.deleted.into_iter());
+    // put them in a set to make them unique
+    let updated_users: HashSet<String> =
+        HashSet::from_iter(updated_chain.map(|device| device.user_id().to_string()));
+    let updated_users_vec = Vec::from_iter(updated_users.iter());
```

Comment:
I think this means the user IDs are passed as `String`s? Which seems inconsistent with other parts of the API, like `register_user_identity_updated_callback` ?

Reply:
Yes, it is.  It's because HashSet (and BTreeSet) didn't like storing `identifiers::UserId`, and it seemed silly to convert it to a string for HashSet, and then convert it to a `UserId`, only for the JavaScript to convert it back to a string.